numpy/_core/src/umath/matmul.c.src

```diff
@@ -515,3 +515,165 @@ NPY_NO_EXPORT void
+ */
+
+/**begin repeat
+ *
+ * #name = CFLOAT, CDOUBLE#
+ * #ctype = npy_cfloat, npy_cdouble#
+ * #type = npy_float, npy_double#
+ * #prefix = c, z#
+ */
+NPY_NO_EXPORT void
+@name@_dotc(char *ip1, npy_intp is1, char *ip2, npy_intp is2,
+            char *op, npy_intp n, void *NPY_UNUSED(ignore))
+{
+#if defined(HAVE_CBLAS)
+    CBLAS_INT is1b = blas_stride(is1, sizeof(@ctype@));
+    CBLAS_INT is2b = blas_stride(is2, sizeof(@ctype@));
+
+    if (is1b && is2b) {
+        double sum[2] = {0., 0.};  /* double for stability */
+
+        while (n > 0) {
+            CBLAS_INT chunk = n < NPY_CBLAS_CHUNK ? n : NPY_CBLAS_CHUNK;
+            @type@ tmp[2];
+
+            CBLAS_FUNC(cblas_@prefix@dotc_sub)(
+                    (CBLAS_INT)chunk, ip1, is1b, ip2, is2b, tmp);
+            sum[0] += (double)tmp[0];
+            sum[1] += (double)tmp[1];
+            /* use char strides here */
+            ip1 += chunk * is1;
+            ip2 += chunk * is2;
+            n -= chunk;
+        }
+        ((@type@ *)op)[0] = (@type@)sum[0];
+        ((@type@ *)op)[1] = (@type@)sum[1];
+    }
+    else
+#endif
+    {
+        double sumr = 0.0;  /* double for stability */
+        double sumi = 0.0;
+        npy_intp i;
+
+        for (i = 0; i < n; i++, ip1 += is1, ip2 += is2) {
+            const @type@ ip1r = ((@type@ *)ip1)[0];
+            const @type@ ip1i = ((@type@ *)ip1)[1];
+            const @type@ ip2r = ((@type@ *)ip2)[0];
+            const @type@ ip2i = ((@type@ *)ip2)[1];
+            /* ip1.conj() * ip2 */
+            sumr += (double)(ip1r * ip2r + ip1i * ip2i);
+            sumi += (double)(ip1r * ip2i - ip1i * ip2r);
+        }
+        ((@type@ *)op)[0] = (@type@)sumr;
+        ((@type@ *)op)[1] = (@type@)sumi;
+    }
+}
+/**end repeat**/
+
+static void
+CLONGDOUBLE_dotc(char *ip1, npy_intp is1, char *ip2, npy_intp is2,
+                 char *op, npy_intp n, void *NPY_UNUSED(ignore))
+{
+    npy_longdouble tmpr = 0.0L;
+    npy_longdouble tmpi = 0.0L;
+    npy_intp i;
+
+    for (i = 0; i < n; i++, ip1 += is1, ip2 += is2) {
+        const npy_longdouble ip1r = ((npy_longdouble *)ip1)[0];
+        const npy_longdouble ip1i = ((npy_longdouble *)ip1)[1];
+        const npy_longdouble ip2r = ((npy_longdouble *)ip2)[0];
+        const npy_longdouble ip2i = ((npy_longdouble *)ip2)[1];
+        /* ip1.conj() * ip2 */
+        tmpr += ip1r * ip2r + ip1i * ip2i;
+        tmpi += ip1r * ip2i - ip1i * ip2r;
+    }
+    ((npy_longdouble *)op)[0] = tmpr;
+    ((npy_longdouble *)op)[1] = tmpi;
+}
+
+static void
+OBJECT_dotc(char *ip1, npy_intp is1, char *ip2, npy_intp is2, char *op, npy_intp n,
+            void *NPY_UNUSED(ignore))
+{
+    npy_intp i;
+    PyObject *tmp1, *tmp2, *tmp = NULL;
+    PyObject **tmp3;
+    for (i = 0; i < n; i++, ip1 += is1, ip2 += is2) {
+        if ((*((PyObject **)ip1) == NULL) || (*((PyObject **)ip2) == NULL)) {
+            tmp1 = Py_False;
+            Py_INCREF(Py_False);
```

Comment:
This comes straight from the loop in `arraytypes` - is it even possible for the pointers to be `NULL`?

Assuming it is, I don't think we can replace by `Py_None`, since adding `None` to anything is not possible. Might it then make sense to just error straight away? At least that would be obvious.

Reply:
But for `matmul` we do the same, of just assigning `Py_None`, so reasonable to do that here too...  Will try that.

Reply:
Yes, and that doesn't matter: `NULL` is synonymous for `None` if it should happen.  Yes, it will error, but explicitly erroring is actually also wrong (because the error will at least mention `None`, while mentioning `NULL` has no meaning in python).

Reply:
Basically, `None` isn't an assumption, it is the correct definition... maybe we should have a `npy_load_object()` helper to do it.  If `arraytypes` does it differently, it is simply buggy...